### PR TITLE
Move session token to redis

### PR DESCRIPTION
**Pull request by udaraweerasinghege**
Fixes https://github.com/mozilla/fxa-auth-server/issues/1044
Ready for review

Sample of what the redis key/val looks like:
```
{"fxa-auth-session740eb35a75bd4908b8fcbe9efd5a29ad":"[{\"tokenId\":\"781d1b50c3114d60b778a605c576cb82eaa1022165a3ce05aaa855e0585325ec\",\"uid\":\"740eb35a75bd4908b8fcbe9efd5a29ad\",\"uaBrowser\":\"Firefox\",\"uaBrowserVersion\":\"54\",\"uaOS\":\"Mac OS X\",\"uaOSVersion\":\"10.12\",\"uaDeviceType\":null,\"lastAccessTime\":1499357068854,\"createdAt\":1499357068854}]"}
```

### CODE REVIEW

File: config/index.js [hunk 1]
```diff
@@ -334,6 +334,32 @@ var conf = convict({
       env: 'SUPPORTED_LANGS'
     }
   },
+  useRedis : {
```
Comment (philbooth) [inline config/index.js]: Apologies for the drive-by on a WIP PR, but maybe move this in with the others and call it something like `redis.enabled`?

File: config/index.js [hunk 2]
```diff
@@ -334,6 +334,32 @@ var conf = convict({
...
+    default: true,
+    doc: 'Enable redis cache',
+    format: Boolean,
+    env: 'USE_REDIS'
+  },
+  redis: {
```
Comment (vladikoff) [inline config/index.js]: @udaraweerasinghege @philbooth we are going to be using redis for `fxa-sessions` and `fxa-devices` (device list cache). I wonder if we need separate redis host configurations or at least support for multiple key prefixes 

cc @jbuck

File: package.json [hunk 3]
```diff
@@ -64,6 +64,7 @@
     "nodemailer": "2.7.2",
     "po2json": "0.4.5",
     "poolee": "1.0.1",
+    "redis": "^2.7.1",
```
Comment (vladikoff) [inline package.json]: remove the `^`

File: bin/key_server.js [hunk 4]
```diff
@@ -74,11 +75,21 @@ function run(config) {
...
+    prefix: config.redis.prefix
+  })
+
   var DB = require('../lib/db')(
     config,
     log,
     Token,
-    UnblockCode
+    UnblockCode,
+    redisClient
```
Comment (vladikoff) [inline bin/key_server.js]: @udaraweerasinghege I think we want to move away from `functional approach` here, AKA requiring stuff in the server.js and then passing it through. Can we just require the redisClient in the files we need it?

File: config/index.js [hunk 5]
```diff
@@ -334,6 +334,32 @@ var conf = convict({
       env: 'SUPPORTED_LANGS'
     }
   },
+  redis: {
+    enabled : {
```
Comment (vladikoff) [inline config/index.js]: style: space between `enabled : `

File: config/index.js [hunk 6]
```diff
@@ -334,6 +334,32 @@ var conf = convict({
...
+      default: true,
+      doc: 'Enable redis cache',
+      format: Boolean,
+      env: 'USE_REDIS'
+    },
+    host: {
+      default: '127.0.0.1',
+      env: 'REDIS_HOST',
+      format: String,
+      doc: 'Url for redis host',
```
Comment (vladikoff) [inline config/index.js]: `Url` -> `url`

File: config/index.js [hunk 7]
```diff
@@ -334,6 +334,32 @@ var conf = convict({
...
+    },
+    port: {
+      default: 6379,
+      env: 'REDIS_PORT',
+      format: 'port',
+      doc: 'port for redis server'
+    },
+    sessionsKeyPrefix: {
+      default: 'fxa-auth-session',
+      env: 'REDIS_KEY_PREFIX',
```
Comment (vladikoff) [inline config/index.js]: `REDIS_KEY_PREFIX` needs to change to reflect that it is the *session* one

File: config/index.js [hunk 8]
```diff
@@ -334,6 +334,32 @@ var conf = convict({
...
+      format: String,
+      doc: 'which key prefix to redis cache'
```
Comment (vladikoff) [inline config/index.js]: `which key prefix to redis cache` needs to change to reflect that this is for the session store , not really a `cache` either

File: lib/db.js [hunk 9]
```diff
@@ -211,8 +220,24 @@ module.exports = (
...
+    if (this.redis) {
+      promises.push(this.redis.getAsync(uid))
+    }
+
     log.trace({ op: 'DB.sessions', uid: uid })
-    return this.pool.get('/account/' + uid + '/sessions')
+    return P.all(promises)
+      .then(results => {
+        const tokens = results[0]
+        const redisTokens = results[1] ? JSON.parse(results[1]) : []
```
Comment (vladikoff) [inline lib/db.js]: can the JSON.parse fail here? do we need guard against issues with that?

File: lib/db.js [hunk 10]
```diff
@@ -349,35 +374,47 @@ module.exports = (
...
-          if (isNotFoundError(err)) {
-            throw error.unknownAccount()
+    if (this.redis) {
+      promises.push(this.redis.getAsync(uid))
+    }
+    return P.all(promises)
+      .then(results => {
+        log.info({op: 'devices hello', name: results})
+        const devices = results[0]
+        const tokens = results[1] ? JSON.parse(results[1]) : []
```
Comment (vladikoff) [inline lib/db.js]: can the JSON.parse fail here? do we need guard against issues with that?

File: lib/db.js [hunk 11]
```diff
@@ -349,35 +374,47 @@ module.exports = (
...
-        },
-        function (err) {
-          if (isNotFoundError(err)) {
-            throw error.unknownAccount()
+    if (this.redis) {
+      promises.push(this.redis.getAsync(uid))
+    }
+    return P.all(promises)
+      .then(results => {
+        log.info({op: 'devices hello', name: results})
```
Comment (vladikoff) [inline lib/db.js]: this log needs to be changed

File: lib/db.js [hunk 12]
```diff
@@ -211,8 +220,24 @@ module.exports = (
...
+    }
+
     log.trace({ op: 'DB.sessions', uid: uid })
-    return this.pool.get('/account/' + uid + '/sessions')
+    return P.all(promises)
+      .then(results => {
+        const tokens = results[0]
+        const redisTokens = results[1] ? JSON.parse(results[1]) : []
+        return tokens.map(function (sessionToken) {
+          const redisToken = redisTokens.find(tok => tok.tokenId === sessionToken.tokenId)
```
Comment (vladikoff) [inline lib/db.js]: let's log here that we are using a redisStore vs dbStore, just like we do in the profile server.

File: lib/db.js [hunk 13]
```diff
@@ -211,8 +220,24 @@ module.exports = (
...
+          return redisToken || sessionToken
```
Comment (vladikoff) [inline lib/db.js]: Is there any info in the sessionToken that is not in the redisToken?

File: lib/db.js [hunk 14]
```diff
@@ -349,35 +374,47 @@ module.exports = (
...
+    if (this.redis) {
+      promises.push(this.redis.getAsync(uid))
+    }
+    return P.all(promises)
+      .then(results => {
+        log.info({op: 'devices hello', name: results})
+        const devices = results[0]
+        const tokens = results[1] ? JSON.parse(results[1]) : []
+
+        return devices.map(item => {
```
Comment (vladikoff) [inline lib/db.js]: can `item` be renamed `device` here?

File: lib/db.js (hunk 10 above)
Comment (vladikoff) [inline lib/db.js]: `tokens` sounds strange, might be confusing for future readers. maybe `redisTokens`, `redisSession`, something like that?

File: lib/db.js [hunk 15]
```diff
@@ -211,8 +220,24 @@ module.exports = (
...
+    ]
+
+    if (this.redis) {
+      promises.push(this.redis.getAsync(uid))
+    }
+
     log.trace({ op: 'DB.sessions', uid: uid })
-    return this.pool.get('/account/' + uid + '/sessions')
+    return P.all(promises)
+      .then(results => {
```
Comment (vbudhram) [inline lib/db.js]: Could use `.spread` here

File: lib/db.js [hunk 16]
```diff
@@ -418,20 +455,37 @@ module.exports = (
...
 
-    return this.pool.post('/sessionToken/' + token.id + '/update', {
-      uaBrowser: token.uaBrowser,
-      uaBrowserVersion: token.uaBrowserVersion,
-      uaOS: token.uaOS,
-      uaOSVersion: token.uaOSVersion,
-      uaDeviceType: token.uaDeviceType,
-      lastAccessTime: token.lastAccessTime
-    })
+    if (this.redis) {
```
Comment (vbudhram) [inline lib/db.js]: Not sure if you need the `if` block here because it get caught in the above `if` block.

File: lib/db.js [hunk 17]
```diff
@@ -418,20 +455,37 @@ module.exports = (
   }
 
   DB.prototype.updateSessionToken = function (token, userAgentString) {
```
Comment (vbudhram) [inline lib/db.js]: Did you guys wanted to remove a session token from Redis when it gets deleted?

File: lib/db.js (hunk 9 above)
Comment (udaraweerasinghege) [inline lib/db.js]: Cannot fail since redis either returns null or what we have previously stored, and we explicitly store `json.stringified` zobjects

File: lib/db.js (hunk 13 above)
Comment (udaraweerasinghege) [inline lib/db.js]: sessionTokens contain the following properties that are not in redisToken, however I can add them if we need. Do we need them?

```
{ 
  deviceId: null,
  deviceName: null,
  deviceType: null,
  deviceCreatedAt: null,
  deviceCallbackURL: null,
  deviceCallbackPublicKey: null,
  deviceCallbackAuthKey: null 
}
```

File: lib/db.js (hunk 17 above)
Comment (udaraweerasinghege) [inline lib/db.js]: cc; @philbooth @vladikoff

Comment (philbooth) [inline lib/db.js]: Yep, we definitely do!

File: lib/db.js (hunk 13 above)
Comment (philbooth) [inline lib/db.js]: Yep, let's make it so the same properties are always returned, regardless of whether the object is from Redis or MySQL. Any properties that aren't in Redis should come from MySQL instead.

File: lib/db.js (hunk 17 above)
Comment (philbooth) [inline lib/db.js]: I hadn't thought about the delete part before, nice one for catching it @vbudhram and sorry for not mentioning it before @udara!

If we want parity with the db, and I think we do, it means session tokens should be removed from redis in each of the following:

* `deleteSessionToken`
* `deleteAccount`
* `resetAccount`

I think that's it, unless there are others I've forgotten about?

File: lib/db.js [hunk 18]
```diff
@@ -529,7 +588,11 @@ module.exports = (
         uid: sessionToken && sessionToken.uid
       }
     )
-    return this.pool.del('/sessionToken/' + sessionToken.id)
+    const promises = [this.pool.del('/sessionToken/' + sessionToken.id)]
+    if (this.redis) {
+      promises.push(this.redis.del(sessionToken.uid))
```
Comment (philbooth) [inline lib/db.js]: Will this delete all of the session tokens for that `uid`? I think we only want to delete an individual session token at this point.

File: test/remote/db_tests.js [hunk 19]
```diff
@@ -156,27 +169,25 @@ describe('remote db', function() {
...
+          assert.equal(redisSetSpy.lastCall, null, 'session token was not updated')
+          return db.sessionToken(tokenId)
         })
         .then(function(sessionToken) {
           sessionToken.lastAccessTime -= TOKEN_FRESHNESS_THRESHOLD
           return db.updateSessionToken(sessionToken, 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10.10; rv:41.0) Gecko/20100101 Firefox/41.0')
         })
-        .then(function() {
+        .then(function(sessionTokens) {
+          assert.ok(redisSetSpy.lastCall, 'session token was updated')
```
Comment (philbooth) [inline test/remote/db_tests.js]: We should assert that the arguments to this call were what we expect them to be.

File: lib/db.js [hunk 20]
```diff
@@ -211,8 +220,29 @@ module.exports = (
...
+    return P.all(promises)
+      .spread((tokens, redisTokens) => {
+        const redisSessionTokens = redisTokens ? JSON.parse(redisTokens) : []
+        return tokens.map(function (sessionToken) {
+          const redisToken = redisSessionTokens.find(tok => tok.tokenId === sessionToken.tokenId)
+          const toLog = redisToken ? 'redis.session' : 'DB.session'
+          const mergedToken = Object.assign({}, sessionToken, redisToken)
+          log.info({
+            op: toLog,
+            name: mergedToken
```
Comment (philbooth) [inline lib/db.js]: `name` seems like a funny name to log `mergedToken` as. Copy-paste artifact?

File: lib/db.js [hunk 21]
```diff
@@ -211,8 +220,29 @@ module.exports = (
...
+
+    log.trace({ op: 'sessions', uid: uid })
+    return P.all(promises)
+      .spread((tokens, redisTokens) => {
+        const redisSessionTokens = redisTokens ? JSON.parse(redisTokens) : []
+        return tokens.map(function (sessionToken) {
+          const redisToken = redisSessionTokens.find(tok => tok.tokenId === sessionToken.tokenId)
+          const toLog = redisToken ? 'redis.session' : 'DB.session'
+          const mergedToken = Object.assign({}, sessionToken, redisToken)
+          log.info({
```
Comment (philbooth) [inline lib/db.js]: Actually, I'm wondering how valuable this log message is. There's going to be quite a lot of them if we're logging for every session in the array each time. I'd suggest getting rid of it, unless there's some specific reason we think we'll need to go looking for it in Kibana one day...?

Comment (philbooth) [inline lib/db.js]: For instance, if we're interested in tracking how many sessions are being returned from redis and how many from MySQL, it would be saner and more efficient to log a single message that summarises that outside of the loop body.

Comment (philbooth) [inline lib/db.js]: (and there's no equivalent log entry for `devices` I notice...)

File: lib/db.js [hunk 22]
```diff
@@ -211,8 +220,29 @@ module.exports = (
...
-    return this.pool.get('/account/' + uid + '/sessions')
+    const promises = [
+      this.pool.get('/account/' + uid + '/sessions')
+    ]
+
+    if (this.redis) {
+      promises.push(this.redis.getAsync(uid))
+    }
+
+    log.trace({ op: 'sessions', uid: uid })
```
Comment (philbooth) [inline lib/db.js]: AFAIK, the point of the `trace` log level is to log function names as they're entered. So I'd keep `op` as `'DB.sessions'`, because that is the object this function was called on.

And for consistency with the other `trace` calls, I'd keep it as first line in the function.

File: lib/db.js [hunk 23]
```diff
@@ -418,20 +458,35 @@ module.exports = (
   }
 
   DB.prototype.updateSessionToken = function (token, userAgentString) {
-    log.trace({ op: 'DB.updateSessionToken', uid: token && token.uid })
-
-    if (! token.update(userAgentString)) {
+    log.trace({ op: 'REDIS.updateSessionToken', uid: token && token.uid })
```
Comment (philbooth) [inline lib/db.js]: Like the other one, I'd keep this trace as `op: 'DB.updateSessionToken'` because that matches the name of the object this method is part of.

File: lib/db.js [hunk 24]
```diff
@@ -418,20 +458,35 @@ module.exports = (
...
       uaBrowserVersion: token.uaBrowserVersion,
       uaOS: token.uaOS,
       uaOSVersion: token.uaOSVersion,
       uaDeviceType: token.uaDeviceType,
-      lastAccessTime: token.lastAccessTime
-    })
+      lastAccessTime: token.lastAccessTime,
+      createdAt: token.createdAt
+    }]
+    let value = []
```
Comment (philbooth) [inline lib/db.js]: I'm well into bike-shedding territory here, but `value` seems like a slightly misleading name for the result array here. `values` or `results` or `sessionTokens` are all better names I reckon.

File: test/remote/db_tests.js [hunk 25]
```diff
@@ -202,6 +213,8 @@ describe('remote db', function() {
           return db.deleteSessionToken(sessionToken)
         })
         .then(function() {
+          assert.equal(redisDelSpy.lastCall.args[0], account.uid)
```
Comment (philbooth) [inline test/remote/db_tests.js]: We need a bit more coverage here around the delete behaviour I think. For instance, if there is a problem with `deleteSessionToken` accidentally removing all tokens for the user, this assertion fails to spot that. If there were two tokens and it asserts the correct one is deleted, then we'd be in a stronger position.

File: lib/db.js [hunk 26]
```diff
@@ -349,35 +379,45 @@ module.exports = (
...
+            name: device.name,
+            type: device.type,
+            pushCallback: device.callbackURL,
+            pushPublicKey: device.callbackPublicKey,
+            pushAuthKey: device.callbackAuthKey,
+            uaBrowser: hasToken ? token.uaBrowser : device.uaBrowser,
+            uaBrowserVersion: hasToken ? token.uaBrowserVersion : device.uaBrowserVersion,
+            uaOS: hasToken ? token.uaOS : device.uaOS,
+            uaOSVersion: hasToken ? token.uaOSVersion : device.uaOSVersion,
+            uaDeviceType: hasToken ? token.uaDeviceType : device.uaDeviceType
```
Comment (philbooth) [inline lib/db.js]: Rather than explicitly checking `hasToken` 5 times, would this code be a little bit nicer if you leaned on the `Object.assign` trick you use in `DB.sessions`?

File: lib/db.js (hunk 21 above)
Comment (udaraweerasinghege) [inline lib/db.js]: The purpose was in fact to track how many sessions are from redis and how many are from sql, oversight on my part to put it in the loop. Will move outside.

Comment (udaraweerasinghege) [inline lib/db.js]: @philbooth What do you think is the best format for this log to track redis/DB hit rate?
```
{redis.sessionToken: 3, db.sessionTokens: 5
```
or 

```
{redis.sessionToken: [redissessionTokens], db.sessionTokens: [dbsessionTokens]
```
or something else altogether. I'm not super aware of how kibana is configured.

File: lib/db.js (hunk 18 above)
Comment (udaraweerasinghege) [inline lib/db.js]: Good catch!

File: lib/db.js (hunk 21 above)
Comment (philbooth) [inline lib/db.js]: Hmmm, I'd prolly go for something like this:

```js
log.info({
  op: 'db.sessions.count',
  mysql: mysqlSessionTokens.length,
  redis: redisSessionTokens.length
})
```

...plus any other properties you think might be worthwhile as extra info.

Comment (udaraweerasinghege) [inline lib/db.js]: Awesome, thanks!

File: lib/db.js [hunk 27]
```diff
@@ -419,19 +461,38 @@ module.exports = (
...
       uaOS: token.uaOS,
       uaOSVersion: token.uaOSVersion,
       uaDeviceType: token.uaDeviceType,
-      lastAccessTime: token.lastAccessTime
-    })
+      lastAccessTime: token.lastAccessTime,
+      createdAt: token.createdAt
+    }]
+    let sessionTokens = []
+    if (features.isLastAccessTimeEnabledForUser(uid)) {
```
Comment (vladikoff) [inline lib/db.js]: @udaraweerasinghege do we need to modify `isLastAccessTimeEnabledForUser` so it's not using email anymore?

cc @philbooth

Comment (udaraweerasinghege) [inline lib/db.js]: The body of the function is
 ```javascript
return lastAccessTimeUpdates.enabled && (
        isSampledUser(lastAccessTimeUpdates.sampleRate, uid, 'lastAccessTimeUpdates') ||
        lastAccessTimeUpdates.enabledEmailAddresses.test(email)
```
so it'll work without using the email. There's a call in db.devices which does use email.

Comment (vladikoff) [inline lib/db.js]: let's remove ` lastAccessTimeUpdates.enabledEmailAddresses.test(email)` and also in devices

Comment (vladikoff) [inline lib/db.js]: UPDATE: so remove `lastAccessTime: marshallLastAccessTime(item.lastAccessTime, uid, item.email),` from devices. Otherwise when we go from 100% to 30% the 70% will have a blank "start syncing" string

File: config/index.js [hunk 28]
```diff
@@ -334,6 +334,32 @@ var conf = convict({
       env: 'SUPPORTED_LANGS'
     }
   },
+  redis: {
+    enabled: {
+      default: true,
```
Comment (vladikoff) [inline config/index.js]: @udaraweerasinghege should we set this to false in dev.json?  OR I wonder if we need to bring back redis to fxa-local-dev as a require dependency, just so all devs test this functionality

File: lib/db.js [hunk 29]
```diff
@@ -349,46 +380,46 @@ module.exports = (
...
-            throw error.unknownAccount()
+    if (this.redis) {
+      promises.push(this.redis.getAsync(uid))
+    }
+    return P.all(promises)
+      .spread((devices, redisTokens) => {
+        const redisSessionTokens = redisTokens ? JSON.parse(redisTokens) : []
+        log.info({
+          op: 'db.devices.redisSessionTokens',
+          hasRedisTokens: !! redisSessionTokens.length
```
Comment (philbooth) [inline lib/db.js]: Maybe this one should also log the length like the one in `sessions` does?

File: lib/features.js [hunk 30]
```diff
@@ -21,10 +21,9 @@ module.exports = config => {
      * @param uid   Buffer or String
      * @param email String
      */
-    isLastAccessTimeEnabledForUser (uid, email) {
+    isLastAccessTimeEnabledForUser (uid) {
       return lastAccessTimeUpdates.enabled && (
-        isSampledUser(lastAccessTimeUpdates.sampleRate, uid, 'lastAccessTimeUpdates') ||
-        lastAccessTimeUpdates.enabledEmailAddresses.test(email)
+        isSampledUser(lastAccessTimeUpdates.sampleRate, uid, 'lastAccessTimeUpdates')
       )
```
Comment (philbooth) [inline lib/features.js]: Let's ditch the wrapping parentheses around the second part of this condition now that they're redundant.

File: config/index.js (hunk 28 above)
Comment (philbooth) [inline config/index.js]: Yeah, if it's straightforward enough, let's make sure this runs by default in local dev. It is (or will be) core functionality for the content server.

**Comment by philbooth** [comment]
@udaraweerasinghege, @vladikoff, this is looking pretty damn splendid to me, I'm officially r+! As soon as you guys are both happy, let's get this rebased, squashed and merged!

File: config/index.js [hunk 31]
```diff
@@ -578,12 +604,6 @@ var conf = convict({
       format: Number,
       default: 1,
```
Comment (vladikoff) [inline config/index.js]: change this to 0.3 , and to 1 in development mode

File: lib/db.js [hunk 32]
```diff
@@ -27,6 +31,11 @@ module.exports = (
 
   function DB(options) {
     this.pool = new Pool(options.url)
+    this.redis = config.redis.enabled ? redis.createClient({
+      host: config.redis.host,
+      port: config.redis.port,
+      prefix: config.redis.sessionsKeyPrefix
+    }) : undefined
```
Comment (vladikoff) [inline lib/db.js]: @udaraweerasinghege can we add a `log.info` after this to show us if redis got enabled or not. 

If it is `on` then log the config values it is using, so we can see this info when server starts

File: lib/db.js [hunk 33]
```diff
@@ -212,7 +221,29 @@ module.exports = (
...
+      this.pool.get('/account/' + uid + '/sessions')
+    ]
+
+    if (this.redis) {
+      promises.push(this.redis.getAsync(uid))
+    }
+
+    return P.all(promises)
+      .spread((mysqlSessionTokens, redisTokens) => {
+        const redisSessionTokens = redisTokens ? JSON.parse(redisTokens) : []
```
Comment (vladikoff) [inline lib/db.js]: Could you please add a short doc string that explains what the 5 lines below are doing. This is for future readers of this code

File: lib/db.js [hunk 34]
```diff
@@ -349,46 +380,46 @@ module.exports = (
...
-        },
-        function (err) {
-          if (isNotFoundError(err)) {
-            throw error.unknownAccount()
+    if (this.redis) {
+      promises.push(this.redis.getAsync(uid))
+    }
+    return P.all(promises)
+      .spread((devices, redisTokens) => {
+        const redisSessionTokens = redisTokens ? JSON.parse(redisTokens) : []
```
Comment (vladikoff) [inline lib/db.js]: Could you please add a short doc string that explains what the 5 lines below are doing. This is for future readers of this code

File: lib/db.js [hunk 35]
```diff
@@ -419,19 +450,38 @@ module.exports = (
...
+      lastAccessTime: token.lastAccessTime,
+      createdAt: token.createdAt
+    }]
+    let sessionTokens = []
+    if (features.isLastAccessTimeEnabledForUser(uid)) {
+      return this.redis.getAsync(uid)
+      .then(res => {
+        if (res) {
+          res = JSON.parse(res)
+          const filteredRes = res.filter(tok => tok.tokenId !== token.id)
```
Comment (vladikoff) [inline lib/db.js]: Could you please add a short doc string that explains what the several lines below are doing. This is for future readers of this code...

**Comment by udaraweerasinghege** [comment]
squashed + rebased, and ready for a final look

File: lib/tokens/session_token.js [hunk 36]
```diff
@@ -72,36 +58,11 @@ module.exports = (log, Token, config) => {
         lastAccessTime: Date.now()
       }, userAgentString, log)
 
-      if (this.isFresh(freshData)) {
-        return false
-      }
-
       this.setUserAgentInfo(freshData)
 
       return true
```
Comment (philbooth) [inline lib/tokens/session_token.js]: Let's ditch this `return true` now that nothing checks the return value.

File: lib/tokens/session_token.js [hunk 37]
```diff
@@ -56,14 +49,7 @@ module.exports = (log, Token, config) => {
...
-    // the session token needs updating.
-    //
-    // If the session token has not changed, allowing up to an hour of
-    // leeway for lastAccessTime, return false.
-    //
-    // Otherwise, update properties on this then return true.
-    //
+    // Parse the user agent string, then update properties on this
     // It is the caller's responsibility to update the database.
     update(userAgentString) {
```
Comment (philbooth) [inline lib/tokens/session_token.js]: Definitely not for now but fwiw, in a future refactoring I suspect we may get rid of this method and just call `userAgent` from the route handler directly. It feels like an unnecessary indirection after these changes.

File: test/remote/db_tests.js [hunk 38]
```diff
@@ -339,21 +374,11 @@ describe('remote db', function() {
...
-            lastAccessTimeUpdates.enabled = true
-            // Delete the device
+
             return db.deleteDevice(account.uid, deviceInfo.id)
-              .catch(function () {
-                assert(false, 'deleting a device should not have failed')
-              })
+            .catch(function () {
+              assert(false, 'deleting a device should not have failed')
+            })
```
Comment (philbooth) [inline test/remote/db_tests.js]: This indentation change seems unnecessary?

**Comment by philbooth** [comment]
> ...ready for a final look

I hereby reaffirm my r+!
